File: web/src/lib/device/services.ts
import { Temporal } from "@js-temporal/polyfill";
import { Engine, type Param } from "./engine";
import * as c from "./contracts";
import { categories, daysBetween } from "../domain";
import { ApiError } from "../errors";
import type { Trip } from "../types";
import { rate } from "./fx";
import { refresh } from "./schema";

const id = () => crypto.randomUUID();
const amount = (n: number) => Math.round((n + Number.EPSILON) * 100) / 100;
const invalid = (message: string): never => {
  throw new ApiError(message, 400);
};
export async function trip(e: Engine, tid: string) {
  const rows = await e.rows<Trip>("select * from dim_trip where trip_id=?", [
    tid,
  ]);
  if (!rows.length) throw new ApiError("找不到這趟旅程。", 404);
  return rows[0];
}
export async function revision(e: Engine, tid: string) {
  await trip(e, tid);
  return (
    await e.rows<{ revision: number }>(
      "select revision from app_trip_revision where trip_id=?",
      [tid],
    )
  )[0].revision;
}
export async function checkRevision(e: Engine, tid: string, expected?: number) {
  if ((await revision(e, tid)) !== expected)
    throw new ApiError(
      "資料已在其他頁面更新。請重新載入後再試，避免覆蓋新的內容。",
      409,
    );
}
async function next(e: Engine, table: string, key: string) {
  return (
    await e.rows<{ id: number }>(
      `select coalesce(max(${key}),0)+1 id from ${table}`,
    )
  )[0].id;
}
async function requireItem(
  e: Engine,
  tid: string,
  table: string,
  key: string,
  value: number,
) {
  if (
    !(
      await e.rows(`select ${key} from ${table} where trip_id=? and ${key}=?`, [
        tid,
        value,
      ])
    ).length
  )
    throw new ApiError("找不到這項資料，可能已被刪除。", 404);
}
async function place(
  e: Engine,
  name: string,
  locality: string,
  googleId: string,
) {
  if (!name) return null;
  const pid = id();
  await e.exec(
    "insert into dim_place(place_id,name,locality,gmaps_place_id) values (?,?,?,?)",
    [pid, name, locality, googleId || null],
  );
  return pid;
}
function duration(start: string, end: string) {
  const n = daysBetween(start, end) + 1;
  if (n < 1 || n > 366) invalid("旅程長度需介於 1 到 366 天。");
  return n;
}
export async function createTrip(e: Engine, body: unknown) {
  const b = c.tripCreate.parse(body);
  duration(b.start, b.end);
  if (new Set(b.members).size !== b.members.length)
    invalid("旅伴名字不可重複。");
  const tid = id();
  await e.exec(
    "insert into dim_trip(trip_id,name,start_date,end_date,dates_label,home_currency,local_currency,budget_home) values (?,?,?,?,?,?,?,?)",
    [
      tid,
      b.name,
      b.start,
      b.end,
      `${b.start} — ${b.end}`,
      b.home,
      b.local,
      amount(b.budget),
    ],
  );
  for (const [i, name] of b.members.entries())
    await e.exec("insert into dim_member values (?,?,?,?)", [
      id(),
      tid,
      name,
      i === 0,
    ]);
  for (const key of Object.keys(categories))
    await e.exec("insert into dim_trip_budget values (?,?,0)", [tid, key]);
  await refresh(e, tid);
  return { trip_id: tid };
}
export async function editTrip(e: Engine, tid: string, body: unknown) {
  const b = c.tripEdit.parse(body),
    old = await trip(e, tid),
    n = duration(b.start, b.end);
  if (
    (
      await e.rows(
        "select 1 from dwd_itinerary_item where trip_id=? and day_no>? limit 1",
        [tid, n],
      )
    ).length
  )
    invalid("縮短旅程前，請先調整超出日期的行程。");
  if (
    (
      await e.rows(
        "select 1 from dwd_expense where trip_id=? and not is_deleted and cast(spent_at as date)>? limit 1",
        [tid, b.end],
      )
    ).length
  )
    invalid("結束日期需包含已記錄的支出日期。");
  const budgets = Object.keys(categories).map(
    (key) => [key, amount(b.category_budgets[key] ?? 0)] as const,
  );
  if (budgets.reduce((sum, [, v]) => sum + v, 0) > amount(b.budget) + 1e-8)
    invalid("分類預算加總超過總預算，請調整後再儲存。");
  await e.exec(
    "update dim_trip set name=?,start_date=?,end_date=?,dates_label=?,local_currency=?,budget_home=? where trip_id=?",
    [
      b.name,
      b.start,
      b.end,
      `${b.start} — ${b.end}`,
      b.local,
      amount(b.budget),
      tid,
    ],
  );
  for (const [key, value] of budgets)
    await e.exec("insert or replace into dim_trip_budget values (?,?,?)", [
      tid,
      key,
      value,
    ]);
  if (old.start_date !== b.start)
    await e.exec(
      "update dwd_expense set day_no=greatest(0,date_diff('day',cast(? as date),cast(spent_at as date))+1) where trip_id=?",
      [b.start, tid],
    );
  await refresh(e, tid);
  return { ok: true };
}
export async function addMember(e: Engine, tid: string, body: unknown) {
  const { name } = c.member.parse(body);
  await trip(e, tid);
  if (
    (
      await e.rows(
        "select 1 from dim_member where trip_id=? and display_name=?",
        [tid, name],
      )
    ).length
  )
    invalid("這位旅伴已經在名單中。");
  await e.exec("insert into dim_member values (?,?,?,false)", [
    id(),
    tid,
    name,
  ]);
  await refresh(e, tid);
  return { ok: true };
}
export async function changeCurrency(e: Engine, tid: string, body: unknown) {
  const { home } = c.currencyChange.parse(body),
    old = await trip(e, tid);
  if (home === old.home_currency) return { ok: true };
  const { factor, date } = await rate(e, old.home_currency, home);
  await e.exec(
    "update dim_trip set home_currency=?,budget_home=budget_home*? where trip_id=?",
    [home, factor, tid],
  );
  await e.exec(
    "update dim_trip_budget set planned_home=planned_home*? where trip_id=?",
    [factor, tid],
  );
  await e.exec(
    "update dwd_expense set booked_home_amount=booked_home_amount*?,booked_home_currency=? where trip_id=?",
    [factor, home, tid],
  );
  await e.exec(
    "insert into dwd_currency_change(change_id,trip_id,from_ccy,to_ccy,factor,rate_date) values (?,?,?,?,?,?)",
    [id(), tid, old.home_currency, home, factor, date],
  );
  await refresh(e, tid);
  return { ok: true };
}
async function voidExpense(e: Engine, tid: string, eid: number) {
  if (
    !(
      await e.rows(
        "select 1 from dwd_expense where trip_id=? and expense_id=? and not is_deleted",
        [tid, eid],
      )
    ).length
  )
    invalid("這筆支出已不存在，請重新載入。");
  await e.exec(
    "update dwd_shopping_item set expense_id=null,is_bought=false,updated_at=now() where expense_id=?",
    [eid],
  );
  await e.exec("update dwd_expense set is_deleted=true where expense_id=?", [
    eid,
  ]);
}
export async function saveExpense(
  e: Engine,
  tid: string,
  body: unknown,
  expected?: number,
) {
  const b = c.expense.parse(body),
    existing = await e.rows<{ expense_id: number }>(
      "select expense_id from dwd_expense where trip_id=? and submission_id=?",
      [tid, b.submission_id],
    );
  if (existing.length) return { expense_id: existing[0].expense_id };
  if (b.replaces != null) await checkRevision(e, tid, expected);
  const t = await trip(e, tid),
    members = (
      await e.rows<{ member_id: string }>(
        "select member_id from dim_member where trip_id=?",
        [tid],
      )
    ).map((m) => m.member_id),
    split = [...new Set(b.split)];
  if (!members.includes(b.payer) || split.some((m) => !members.includes(m)))
    invalid("付款人與分攤旅伴必須來自這趟旅程。");
  if (b.when > t.end_date) invalid("付款日期不能晚於旅程結束日期。");
  if (b.shopping_id != null) {
    const item = await e.rows<{ expense_id: number | null }>(
      "select expense_id from dwd_shopping_item where trip_id=? and item_id=?",
      [tid, b.shopping_id],
    );
    if (
      !item.length ||
      (item[0].expense_id !== null && item[0].expense_id !== b.replaces)
    )
      invalid("這件物品已經記過帳，請到支出紀錄確認。");
  }
  const { factor, date } = await rate(e, b.currency, t.home_currency),
    eid = await next(e, "dwd_expense", "expense_id"),
    paid = amount(b.amount);
  if (b.replaces != null) await voidExpense(e, tid, b.replaces);
  await e.exec(
    `insert into dwd_expense(expense_id,trip_id,day_no,spent_at,title,category,amount,currency,member_id,fx_rate_date,booked_home_amount,booked_home_currency,fx_home_currency,applied_fx_rate,submission_id) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)`,
    [
      eid,
      tid,
      Math.max(0, daysBetween(t.start_date, b.when) + 1),
      `${b.when} 12:00:00`,
      b.title,
      b.category,
      paid,
      b.currency,
      b.payer,
      date,
      paid * factor,
      t.home_currency,
      t.home_currency,
      factor,
      b.submission_id,
    ],
  );
  for (const m of split)
    await e.exec("insert into dwd_expense_split values (?,?,?)", [
      eid,
      m,
      1 / split.length,
    ]);
  if (b.shopping_id != null)
    await e.exec(
      "update dwd_shopping_item set expense_id=?,is_bought=true,updated_at=now() where item_id=?",
      [eid, b.shopping_id],
    );
  await refresh(e, tid);
  return { expense_id: eid };
}
export async function saveActivity(
  e: Engine,
  tid: string,
  body: unknown,
  itemId?: number,
) {
  const b = c.activity.parse(body),
    t = await trip(e, tid);
  if (b.day > t.n_days) invalid("行程日期超出旅程範圍。");
  if (itemId !== undefined)
    await requireItem(e, tid, "dwd_itinerary_item", "item_id", itemId);
  const pid = await place(e, b.place, b.locality, b.google_id),
    values: Param[] = [
      b.day,
      b.time.slice(0, 5),
      b.kind,
      b.title,
      pid,
      amount(b.cost),
      b.currency,
      b.notes,
    ];
  if (itemId === undefined)
    await e.exec(
      "insert into dwd_itinerary_item values (?,?,?,?,?,?,?,?,?,?)",
      [await next(e, "dwd_itinerary_item", "item_id"), tid, ...values],
    );
  else
    await e.exec(
      "update dwd_itinerary_item set day_no=?,start_time=?,kind=?,title=?,place_id=?,planned_cost=?,planned_ccy=?,notes=? where trip_id=? and item_id=?",
      [...values, tid, itemId],
    );
  await refresh(e, tid);
  return { ok: true };
}
export function utcTime(local: string, zone: string) {
  return Temporal.PlainDateTime.from(local).toZonedDateTime(zone, {
    disambiguation: "reject",
  }).epochMilliseconds;
}
export async function saveBooking(
  e: Engine,
  tid: string,
  body: unknown,
  bid?: number,
) {
  const b = c.booking.parse(body);
  await trip(e, tid);
  let start: number, end: number;
  try {
    start = utcTime(b.start, b.start_zone);
    end = utcTime(b.end, b.end_zone);
  } catch {
    invalid(
      "請確認時區與當地時間；夏令時間切換造成的不存在或重複時間，請改用明確時間。",
    );
  }
  if (end! < start!)
    invalid("請確認時區與起訖時間：結束時間不能早於開始時間。");
  if (bid !== undefined)
    await requireItem(e, tid, "dwd_booking", "booking_id", bid);
  const pid = await place(e, b.place, b.locality, b.google_id),
    values: Param[] = [
      b.kind,
      b.title,
      b.provider,
      b.ref,
      b.confirmation,
      b.origin,
      b.destination,
      b.start,
      b.end,
      pid,
      b.price === null ? null : amount(b.price),
      b.currency,
      b.notes,
      b.start_zone,
      b.end_zone,
    ];
  if (bid === undefined)
    await e.exec(
      "insert into dwd_booking(booking_id,trip_id,kind,title,provider,ref_code,confirmation,origin,destination,starts_at,ends_at,place_id,price,price_ccy,notes,start_zone,end_zone) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
      [await next(e, "dwd_booking", "booking_id"), tid, ...values],
    );
  else
    await e.exec(
      "update dwd_booking set kind=?,title=?,provider=?,ref_code=?,confirmation=?,origin=?,destination=?,starts_at=?,ends_at=?,place_id=?,price=?,price_ccy=?,notes=?,start_zone=?,end_zone=? where trip_id=? and booking_id=?",
      [...values, tid, bid],
    );
  await refresh(e, tid);
  return { ok: true };
}
export async function saveShopping(
  e: Engine,
  tid: string,
  body: unknown,
  itemId?: number,
) {
  const b = c.shopping.parse(body);
  await trip(e, tid);
  const values: Param[] = [b.title, b.where, amount(b.price), b.currency];
  if (itemId === undefined) {
    let value = (
      await e.rows<{ id: number }>("select nextval('seq_shopping') id")
    )[0].id;
    while (
      (await e.rows("select 1 from dwd_shopping_item where item_id=?", [value]))
        .length
    )
      value = (
        await e.rows<{ id: number }>("select nextval('seq_shopping') id")
      )[0].id;
    await e.exec(
      "insert into dwd_shopping_item(item_id,trip_id,title,where_hint,planned_price,planned_ccy) values (?,?,?,?,?,?)",
      [value, tid, ...values],
    );
  } else {
    await requireItem(e, tid, "dwd_shopping_item", "item_id", itemId);
    await e.exec(
      "update dwd_shopping_item set title=?,where_hint=?,planned_price=?,planned_ccy=?,updated_at=now() where trip_id=? and item_id=?",
      [...values, tid, itemId],
    );
  }
  await refresh(e, tid);
  return { ok: true };
}
export async function bought(
  e: Engine,
  tid: string,
  itemId: number,
  body: unknown,
) {
  const b = c.bought.parse(body);
  await requireItem(e, tid, "dwd_shopping_item", "item_id", itemId);
  const [item] = await e.rows<{ expense_id: number | null }>(
    "select expense_id from dwd_shopping_item where item_id=?",
    [itemId],
  );
  if (item.expense_id !== null && !b.bought)
    invalid("請先刪除對應支出，再取消已購買。");
  await e.exec(
    "update dwd_shopping_item set is_bought=?,updated_at=now() where item_id=?",
    [b.bought, itemId],
  );
  await refresh(e, tid);
  return { ok: true };
}
export async function remove(
  e: Engine,
  tid: string,
  kind: string,
  itemId: number,
) {
  if (kind === "expenses") await voidExpense(e, tid, itemId);
  else {
    const mapping: Record<string, [string, string]> = {
      activities: ["dwd_itinerary_item", "item_id"],
      bookings: ["dwd_booking", "booking_id"],
      shopping: ["dwd_shopping_item", "item_id"],
    };
    if (!mapping[kind]) throw new ApiError("找不到這項資料。", 404);
    const [table, key] = mapping[kind];
    await requireItem(e, tid, table, key, itemId);
    await e.exec(`delete from ${table} where trip_id=? and ${key}=?`, [
      tid,
      itemId,
    ]);
  }
  await refresh(e, tid);
  return { ok: true };
}

File: core/services.py
"""Validated writes: detail records and summaries commit together."""
import datetime as dt
from decimal import Decimal
from uuid import uuid4
from zoneinfo import ZoneInfo
from core import db, fx

CATEGORIES = ["Transport", "Stay", "Food", "Sights", "Shopping", "Other"]
KINDS = ["flight", "transport", "hotel", "sight", "food", "shop"]

def required(value, label):
    value = str(value).strip()
    if not value:
        raise ValueError(f"{label} is required.")
    return value

def money(value, positive=False):
    value = fx.decimal(value)
    if value < 0 or value > Decimal("999999999999"):
        raise ValueError("Amount must be positive." if positive else "Amount must be non-negative and within range.")
    value = value.quantize(Decimal("0.01"))
    if positive and value == 0:
        raise ValueError("Amount must be at least 0.01.")
    return value

def trip(con, tid):
    row = db.q(con,"select * from dim_trip where trip_id=?",[tid])
    if row.empty:
        raise ValueError("Trip not found.")
    return row.iloc[0]

def next_id(con, table, column):
    return con.execute(f"select coalesce(max({column}),0)+1 from {table}").fetchone()[0]

def create_trip(con, name, start, end, home, local, budget, members):
    name = required(name,"Trip name")
    names = [n.strip() for n in members if n.strip()]
    if not names or len(names) != len(set(names)):
        raise ValueError("Enter at least one member; member names must be unique.")
    if end < start or (end-start).days > 365:
        raise ValueError("Trip duration must be between 1 and 366 days.")
    if home not in fx.HOME_CURRENCIES or local not in fx.SPEND_CURRENCIES:
        raise ValueError("Unsupported currency.")
    tid = uuid4().hex
    with db.transaction(con):
        con.execute("""insert into dim_trip(trip_id,name,start_date,end_date,dates_label,home_currency,local_currency,budget_home)
            values (?,?,?,?,?,?,?,?)""",[tid,name,start,end,f"{start:%b %d} – {end:%b %d, %Y}",home,local,money(budget)])
        con.executemany("insert into dim_member values (?,?,?,?)",[(uuid4().hex,tid,n,i==0) for i,n in enumerate(names)])
        con.executemany("insert into dim_trip_budget values (?,?,?)",[(tid,c,0) for c in CATEGORIES])
        db._refresh_dws(con,tid)
    return tid

def update_trip(con, tid, name, start, end, local, budget, category_budgets):
    with db.transaction(con):
        old = trip(con,tid)
        if end < start or (end-start).days > 365:
            raise ValueError("Trip duration must be between 1 and 366 days.")
        if local not in fx.SPEND_CURRENCIES:
            raise ValueError("Unsupported currency.")
        if con.execute("select 1 from dwd_itinerary_item where trip_id=? and day_no>? limit 1",[tid,(end-start).days+1]).fetchone():
            raise ValueError("Move itinerary items within the new trip length first.")
        if con.execute("select 1 from dwd_expense where trip_id=? and not is_deleted and cast(spent_at as date)>? limit 1",[tid,end]).fetchone():
            raise ValueError("Trip end must include all recorded expenses.")
        budget = money(budget)
        amounts = {c:money(category_budgets.get(c,0)) for c in CATEGORIES}
        if sum(amounts.values()) > budget:
            raise ValueError("Category budgets cannot exceed the total budget.")
        con.execute("update dim_trip set name=?,start_date=?,end_date=?,dates_label=?,local_currency=?,budget_home=? where trip_id=?",
            [required(name,"Trip name"),start,end,f"{start:%b %d} – {end:%b %d, %Y}",local,budget,tid])
        for c,amount in amounts.items():
            con.execute("insert or replace into dim_trip_budget values (?,?,?)",[tid,c,amount])
        if start != old.start_date:
            con.execute("update dwd_expense set day_no=greatest(0,date_diff('day',?,cast(spent_at as date))+1) where trip_id=?",[start,tid])
        db._refresh_dws(con,tid)

def add_member(con, tid, name):
    with db.transaction(con):
        trip(con,tid)
        name = required(name,"Member name")
        if con.execute("select 1 from dim_member where trip_id=? and display_name=?",[tid,name]).fetchone():
            raise ValueError("Member name already exists.")
        con.execute("insert into dim_member values (?,?,?,false)",[uuid4().hex,tid,name])
        db._refresh_dws(con,tid)

def change_home_currency(con, tid, home):
    with db.transaction(con):
        old = trip(con,tid).home_currency
        if home not in fx.HOME_CURRENCIES:
            raise ValueError("Unsupported home currency.")
        if home == old:
            return
        factor,date = fx.factor(con,old,home)
        con.execute("update dim_trip set home_currency=?,budget_home=budget_home*? where trip_id=?",[home,factor,tid])
        con.execute("update dim_trip_budget set planned_home=planned_home*? where trip_id=?",[factor,tid])
        con.execute("update dwd_expense set booked_home_amount=booked_home_amount*?,booked_home_currency=? where trip_id=?",[factor,home,tid])
        con.execute("insert into dwd_currency_change(change_id,trip_id,from_ccy,to_ccy,factor,rate_date) values (?,?,?,?,?,?)",[uuid4().hex,tid,old,home,factor,date])
        db._refresh_dws(con,tid)

def save_expense(con, tid, title, category, amount, currency, when, payer, split, submission_id, shopping_id=None, replaces=None):
    with db.transaction(con):
        submission_id = required(submission_id,"Submission ID")
        existing = con.execute("select expense_id from dwd_expense where submission_id=? and trip_id=?",[submission_id,tid]).fetchone()
        if existing:
            return existing[0]
        t = trip(con,tid)
        amount = money(amount,positive=True)
        if currency not in fx.SPEND_CURRENCIES or category not in CATEGORIES:
            raise ValueError("Choose a supported category and currency.")
        members = {m for (m,) in con.execute("select member_id from dim_member where trip_id=?",[tid]).fetchall()}
        split = list(dict.fromkeys(split))
        if payer not in members or not split or not set(split) <= members:
            raise ValueError("Choose a payer and at least one participant from this trip.")
        if when > t.end_date:
            raise ValueError("Expense date must be on or before the trip end date.")
        if shopping_id is not None:
            item = con.execute("select expense_id from dwd_shopping_item where trip_id=? and item_id=?",[tid,shopping_id]).fetchone()
            if item is None or (item[0] is not None and item[0] != replaces):
                raise ValueError("Shopping item is missing or already linked to an expense.")
        rate,date = fx.factor(con,currency,t.home_currency)
        eid = next_id(con,"dwd_expense","expense_id")
        if replaces is not None:
            _void_expense(con,tid,replaces)
        con.execute("""insert into dwd_expense(expense_id,trip_id,day_no,spent_at,title,category,amount,currency,member_id,
            fx_rate_date,booked_home_amount,booked_home_currency,fx_home_currency,applied_fx_rate,submission_id)
            values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)""",[eid,tid,max(0,(when-t.start_date).days+1),dt.datetime.combine(when,dt.time(12)),
            required(title,"Expense description"),category,amount,currency,payer,date,amount*rate,t.home_currency,t.home_currency,rate,submission_id])
        con.executemany("insert into dwd_expense_split values (?,?,?)",[(eid,m,1/len(split)) for m in split])
        if shopping_id is not None:
            con.execute("update dwd_shopping_item set expense_id=?,is_bought=true,updated_at=now() where item_id=?",[eid,shopping_id])
        db._refresh_dws(con,tid)
        return eid

def _void_expense(con, tid, eid):
    if not con.execute("select 1 from dwd_expense where expense_id=? and trip_id=? and not is_deleted",[eid,tid]).fetchone():
        raise ValueError("Expense no longer exists. Reload this page.")
    con.execute("update dwd_shopping_item set expense_id=null,is_bought=false,updated_at=now() where expense_id=?",[eid])
    con.execute("update dwd_expense set is_deleted=true where expense_id=?",[eid])

def delete_expense(con, tid, eid):
    with db.transaction(con):
        _void_expense(con,tid,eid)
        db._refresh_dws(con,tid)

def _place(con, name, locality, google_id):
    if not name.strip():
        return None
    pid = uuid4().hex
    con.execute("insert into dim_place(place_id,name,locality,gmaps_place_id) values (?,?,?,?)",[pid,name.strip(),locality.strip(),google_id.strip() or None])
    return pid

def save_itinerary(con, tid, day, time, kind, title, place, locality, google_id, cost, currency, notes, item_id=None):
    with db.transaction(con):
        t = trip(con,tid)
        if not 1 <= day <= t.n_days or kind not in KINDS or currency not in fx.SPEND_CURRENCIES:
            raise ValueError("Invalid day, activity type or currency.")
        pid = _place(con,place,locality,google_id)
        values = [day,time.strftime("%H:%M"),kind,required(title,"Activity"),pid,money(cost),currency,notes]
        if item_id is None:
            item_id = next_id(con,"dwd_itinerary_item","item_id")
            con.execute("insert into dwd_itinerary_item values (?,?,?,?,?,?,?,?,?,?)",[item_id,tid]+values)
        else:
            con.execute("update dwd_itinerary_item set day_no=?,start_time=?,kind=?,title=?,place_id=?,planned_cost=?,planned_ccy=?,notes=? where trip_id=? and item_id=?",values+[tid,item_id])
        db._refresh_dws(con,tid)

def save_booking(con, tid, kind, title, provider, ref, confirmation, origin, destination, start, end, start_zone, end_zone, place, locality, google_id, price, currency, notes, booking_id=None):
    with db.transaction(con):
        trip(con,tid)
        if kind not in ("flight","hotel","reservation") or currency not in fx.SPEND_CURRENCIES:
            raise ValueError("Invalid booking type or currency.")
        try:
            utc_start = start.replace(tzinfo=ZoneInfo(start_zone)).astimezone(dt.timezone.utc)
            utc_end = end.replace(tzinfo=ZoneInfo(end_zone)).astimezone(dt.timezone.utc)
        except (KeyError,ValueError):
            raise ValueError("Use valid IANA time zones, such as Asia/Tokyo.") from None
        if utc_end < utc_start:
            raise ValueError("End time must be after start time, accounting for time zones.")
        pid = _place(con,place,locality,google_id)
        values = [kind,required(title,"Booking title"),provider,ref,confirmation,origin,destination,start,end,pid,
            None if price is None else money(price),currency,notes,start_zone,end_zone]
        if booking_id is None:
            booking_id = next_id(con,"dwd_booking","booking_id")
            con.execute("""insert into dwd_booking(booking_id,trip_id,kind,title,provider,ref_code,confirmation,origin,destination,
                starts_at,ends_at,place_id,price,price_ccy,notes,start_zone,end_zone) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)""",[booking_id,tid]+values)
        else:
            con.execute("""update dwd_booking set kind=?,title=?,provider=?,ref_code=?,confirmation=?,origin=?,destination=?,starts_at=?,
                ends_at=?,place_id=?,price=?,price_ccy=?,notes=?,start_zone=?,end_zone=? where trip_id=? and booking_id=?""",values+[tid,booking_id])
        db._refresh_dws(con,tid)

def save_shopping(con, tid, title, where, price, currency, item_id=None):
    with db.transaction(con):
        trip(con,tid)
        if currency not in fx.SPEND_CURRENCIES:
            raise ValueError("Unsupported currency.")
        values = [required(title,"Item"),where,money(price),currency]
        if item_id is None:
            con.execute("insert into dwd_shopping_item(trip_id,title,where_hint,planned_price,planned_ccy) values (?,?,?,?,?)",[tid]+values)
        else:
            con.execute("update dwd_shopping_item set title=?,where_hint=?,planned_price=?,planned_ccy=?,updated_at=now() where trip_id=? and item_id=?",values+[tid,item_id])
        db._refresh_dws(con,tid)

def set_bought(con, tid, item_id, bought):
    with db.transaction(con):
        row = con.execute("select expense_id from dwd_shopping_item where trip_id=? and item_id=?",[tid,item_id]).fetchone()
        if row is None:
            raise ValueError("Shopping item not found.")
        if row[0] is not None and not bought:
            raise ValueError("Delete the linked expense before marking this item as unbought.")
        con.execute("update dwd_shopping_item set is_bought=?,updated_at=now() where trip_id=? and item_id=?",[bought,tid,item_id])
        db._refresh_dws(con,tid)

def delete_item(con, tid, table, key, value):
    if (table,key) not in (("dwd_itinerary_item","item_id"),("dwd_booking","booking_id"),("dwd_shopping_item","item_id")):
        raise ValueError("Unsupported deletion.")
    with db.transaction(con):
        con.execute(f"delete from {table} where trip_id=? and {key}=?",[tid,value])
        db._refresh_dws(con,tid)
